### ISSUE TITLE
Setting max juice on transactions

### ISSUE THREAD

**Issue by helins** [Convex-Dev/convex#300]
Reminder: https://github.com/Convex-Dev/convex/blob/34279b17ccd4659db8585f04a090aea8603818a7/convex-core/src/main/java/convex/core/transactions/Invoke.java#L135-L136

Presumably, `.getMaxJuice()` and missing `.setMaxJuice()`, should be implemented on `ATransaction`?

**Comment by helins** [comment]
Or rather putting a max juice argument in `ATransaction` constructor and a `.withMaxJuice()` method.

**Comment by mikera** [comment]
Probably we want to let the client specify an (optional) max juice on a per-transaction basis, in which case it would make sense to be a field on the transaction

**Comment by helins** [comment]
So, following our discussion earlier, all fields on transactions would be optional so that the client can set them as needed before signing (but the data cell on `Invoke` for instance). Otherwise it is a bit absurd that you have to create an `Invoke` object with an address, a sequence, and max juice only for those to be overwritten by the client.

**Comment by mikera** [comment]
Yes, I think the client should usually be able to set this. The `withXXX` pattern is probably fine here.

if you have a crack at this, it will change transaction encodings (breaking db change) and it is also good to think about transaction encoding size, i.e. we want to avoid transmitting any bytes for fields that are not set differently from their default values.
